File: step_16_dynamic_programming/16.2_1d_dp/IB160_longest_valid_parentheses/solution.py
"""
Problem: Longest Valid Parentheses
Difficulty: HARD | XP: 50
Source: InterviewBit

Given a string of '(' and ')', find the length of the longest
valid (well-formed) parentheses substring.
"""
from typing import List


# ============================================================
# APPROACH 1: BRUTE FORCE - Check every even-length substring
# Time: O(n^3)  |  Space: O(1)
# ============================================================
def brute_force(s: str) -> int:
    def is_valid(lo, hi):
        cnt = 0
        for i in range(lo, hi + 1):
            if s[i] == '(':
                cnt += 1
            else:
                cnt -= 1
            if cnt < 0:
                return False
        return cnt == 0

    n = len(s)
    max_len = 0
    for i in range(n):
        for j in range(i + 1, n, 2):  # only even-length substrings
            if is_valid(i, j):
                max_len = max(max_len, j - i + 1)
    return max_len


# ============================================================
# APPROACH 2: OPTIMAL - Stack-based O(n)
# Time: O(n)  |  Space: O(n)
# ============================================================
# Use a stack with a sentinel at -1. Push '(' indices.
# For ')', pop:
#   - Stack empty after pop: push current index as new base.
#   - Otherwise: length = current index - stack top.
# ============================================================
def optimal(s: str) -> int:
    stack = [-1]  # sentinel
    max_len = 0

    for i, ch in enumerate(s):
        if ch == '(':
            stack.append(i)
        else:
            stack.pop()
            if not stack:
                stack.append(i)
            else:
                max_len = max(max_len, i - stack[-1])
    return max_len


# ============================================================
# APPROACH 3: BEST - Two-pass left-right counters (O(1) space)
# Time: O(n)  |  Space: O(1)
# ============================================================
# Left-to-right: count ( and ). Equal => valid length. close>open => reset.
# Right-to-left: handles cases like "((()" where L-R misses.
# Together they cover all cases in two linear passes.
# Real-life use: real-time syntax checking in code editors,
# streaming HTML/XML tag balance validation.
# ============================================================
def best(s: str) -> int:
    max_len = open_cnt = close_cnt = 0

    for ch in s:
        if ch == '(':
            open_cnt += 1
        else:
            close_cnt += 1
        if open_cnt == close_cnt:
            max_len = max(max_len, 2 * close_cnt)
        elif close_cnt > open_cnt:
            open_cnt = close_cnt = 0

    open_cnt = close_cnt = 0
    for ch in reversed(s):
        if ch == '(':
            open_cnt += 1
        else:
            close_cnt += 1
        if open_cnt == close_cnt:
            max_len = max(max_len, 2 * open_cnt)
        elif open_cnt > close_cnt:
            open_cnt = close_cnt = 0

    return max_len


if __name__ == "__main__":
    print("=== Longest Valid Parentheses ===")

    tests   = ["(()",  ")()())", "",  "()", "(()(",  "))(",  "(()((()("]
    expects = [2,      4,        0,   2,    2,       0,      2]

    for s, exp in zip(tests, expects):
        b, op, bst = brute_force(s), optimal(s), best(s)
        ok = b == op == bst == exp
        print(f"s={s!r:15} expect={exp}  b={b} op={op} bst={bst}  {'OK' if ok else 'FAIL'}")

File: step_16_dynamic_programming/16.2_1d_dp/IB160_longest_valid_parentheses/Solution.java
import java.util.*;

// ============================================================
// APPROACH 1: BRUTE FORCE - Check every substring
// Time: O(n^3)  |  Space: O(n)
// ============================================================
// For every pair (i,j), check if substring s[i..j] is a valid
// parentheses string using a counter. Track maximum length.
// ============================================================

class BruteForce {
    private boolean isValid(String s, int lo, int hi) {
        int cnt = 0;
        for (int i = lo; i <= hi; i++) {
            if (s.charAt(i) == '(') cnt++;
            else cnt--;
            if (cnt < 0) return false;
        }
        return cnt == 0;
    }

    public int longestValidParentheses(String s) {
        int n = s.length(), maxLen = 0;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j += 2) { // valid must be even length
                if (isValid(s, i, j)) maxLen = Math.max(maxLen, j - i + 1);
            }
        }
        return maxLen;
    }
}

// ============================================================
// APPROACH 2: OPTIMAL - Stack-based
// Time: O(n)  |  Space: O(n)
// ============================================================
// Push index of '(' onto stack. When ')' is seen:
//   - If stack is empty (or top is also ')'), push current index
//     as the new "base" marker.
//   - Else pop the matching '(' and compute length:
//     current index - stack top (the unmatched index before this group).
// Initialize stack with sentinel -1.
// ============================================================

class Optimal {
    public int longestValidParentheses(String s) {
        int maxLen = 0;
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(-1); // sentinel

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            } else {
                stack.pop();
                if (stack.isEmpty()) {
                    stack.push(i); // new base
                } else {
                    maxLen = Math.max(maxLen, i - stack.peek());
                }
            }
        }
        return maxLen;
    }
}

// ============================================================
// APPROACH 3: BEST - Two-pass left-right counters (O(1) space)
// Time: O(n)  |  Space: O(1)
// ============================================================
// Left-to-right pass: count open/close. When they are equal,
// record length. When close > open, reset both to 0.
// Right-to-left pass: reverse roles (when open > close, reset).
// This handles cases the L-R pass misses (e.g. "(()" - won't
// reset R-L pass catches it by going right-to-left).
// Real-life use: validating code blocks in real-time IDEs,
// parsing nested XML/HTML tags in streaming parsers.
// ============================================================

class Best {
    public int longestValidParentheses(String s) {
        int n = s.length(), maxLen = 0;
        int open = 0, close = 0;

        // Left to right
        for (int i = 0; i < n; i++) {
            if (s.charAt(i) == '(') open++;
            else close++;
            if (open == close) maxLen = Math.max(maxLen, 2 * close);
            else if (close > open) { open = 0; close = 0; }
        }

        open = 0; close = 0;
        // Right to left
        for (int i = n - 1; i >= 0; i--) {
            if (s.charAt(i) == '(') open++;
            else close++;
            if (open == close) maxLen = Math.max(maxLen, 2 * open);
            else if (open > close) { open = 0; close = 0; }
        }

        return maxLen;
    }
}

// ============================================================
// TEST HARNESS
// ============================================================

public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Longest Valid Parentheses ===");

        String[] tests = {"(()", ")()())", "", "()", "(()(", ")((", "(()(((()"};
        int[] expects   = {2,     4,      0,  2,  2,    0,   2};

        for (int t = 0; t < tests.length; t++) {
            String s = tests[t];
            int b  = new BruteForce().longestValidParentheses(s);
            int op = new Optimal().longestValidParentheses(s);
            int bst = new Best().longestValidParentheses(s);
            boolean ok = b == op && op == bst && bst == expects[t];
            System.out.printf("s=%-12s expect=%d  b=%d op=%d bst=%d  %s%n",
                "\""+s+"\"", expects[t], b, op, bst, ok?"OK":"FAIL");
        }
    }
}
